feat: refactor toast

tanvibhakta opened the pull request:
## Description


Zeplin: https://zpl.io/g8yZ0RZ

Fixes https://github.com/appsmithorg/appsmith/issues/19331
Depends on https://github.com/appsmithorg/design-system/pull/318, https://github.com/appsmithorg/design-system/pull/321, https://github.com/appsmithorg/design-system/pull/334

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?

- Manual on storybook 

### Test Plan
https://github.com/appsmithorg/TestSmith/issues/2162

### Issues raised during DP testing
https://github.com/appsmithorg/design-system/pull/333#issuecomment-1425582845
https://github.com/appsmithorg/design-system/pull/333#issuecomment-1426980565

## Checklist:
### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes


### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

albinAppsmith commented on packages/design-system/src/Toast/Toast.stories.tsx:
```diff
@@ -0,0 +1,46 @@
+import { ToastProps } from "./Toast.types";
+
+export default {
+  title: "Design System V2/Toast",
```
Design System

albinAppsmith commented on packages/design-system/src/Toast/Toast.stories.tsx:
```diff
@@ -0,0 +1,46 @@
+  );
+};
+
+export const ToastExample = Template.bind({});
```
Name it like ToastStory instead of example

albinAppsmith commented on packages/design-system/src/Toast/Toast.styles.tsx:
```diff
@@ -0,0 +1,75 @@
+`;
+
+// Variant style definitions
+const YOUR_VARIANTS = css`
```
Remove this if not required

albinAppsmith commented on packages/design-system/src/Toast/Toast.styles.tsx:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  .Toastify__toast-icon {
+    margin-right: 0.5rem;
```
Take it from variables.css...

albinAppsmith commented on packages/design-system/src/Toast/Toast.tsx:
```diff
@@ -0,0 +1,76 @@
+  return (
+    <StyledToast
+      autoClose={5000}
+      className="t--toast-action"
```
Why this class is hardcoded? `t--` is used to denote test classnames. For ads components used ads--

tanvibhakta replied:
don't we need to add the cypress test names to our components to ensure they don't break existing tests?

albinAppsmith replied:
That should be passed from main repo

albinAppsmith replied:
We don't keep any those kind of classnames in DS

albinAppsmith commented on yarn.lock:
```diff
@@ -16229,6 +16209,13 @@ react-tabs@3.1.1:
     clsx "^1.1.0"
     prop-types "^15.5.0"
 
+react-toastify@9.0.3:
```
Where is the package.json file?

albinAppsmith commented:
Vercel deployment is failing here. Is the local build is happening?

albinAppsmith commented on packages/design-system/src/Toast/Toast.tsx:
```diff
@@ -0,0 +1,113 @@
+        <Icon
+          color="var(--ads-v2-color-fg-success)"
+          name="checkbox-circle-fill"
+          size="16px"
```
size is not numbers...

albinAppsmith replied:
Have you taken pull from release?

tanvibhakta replied:
Good catch, I'll fix this!

shadabbuchh commented:
### Bugs in this PR:
**Default Toast:**

1. Functional:

- No neutral option given in the controls

2. Design:

- Wrong border colour
- No Box Shadow
- Wrong colour for the error, warning & success icons
- Wrong height of the warning icon

**Toast with action:**

1. Functional:

- No neutral option given in the controls
- Entering a long text in the content breaks the UI.
<img width="346" alt="Screenshot 2023-02-10 at 3 53 55 PM" src="https://user-images.githubusercontent.com/39921438/218068012-86d4b901-09fc-4702-a186-a6b1e6c3c9e8.png">

2. Design:

- Wrong border colour
- No Box Shadow
- Icons not aligned with the text
<img width="144" alt="Screenshot 2023-02-10 at 3 55 33 PM" src="https://user-images.githubusercontent.com/39921438/218068354-3d9217eb-ddca-4a07-babd-490f0f0ac797.png">

- Wrong colour for the error, warning & success icons
- Wrong height of the warning icon

tanvibhakta commented:
## Default

>   No neutral option given in the controls

This is to be consistent with the developer experience - when we want a regular toast, we don't pass anything through. When we want a special kind of toast, we think of using the `kind` prop.


### Design:

>    Wrong border colour

The value in zeplin seems to be updated now.

>     No Box Shadow

updated

>    Wrong colour for the error, warning & success icons

updated in https://github.com/appsmithorg/design-system/pull/372 

>    Wrong height of the warning icon

It looks like all the icon sizes have now moved from md (16px) to lg (20px) so I've updated all the values accordingly.


## Toast with action

> Entering a long text in the content breaks the UI.

As mentioned in this PR, the toast depends on [the button component](https://github.com/appsmithorg/design-system/pull/334). @albinAppsmith Can you please add a minimum width and height to the button so that this issue does not occur? 

### Design 

> Wrong border colour
> No Box Shadow
> Wrong colour for the error, warning & success icons

Fixed above

> Icons not aligned with the text

It looks like the update to size has fixed this - let me know? 

> The tick mark in the success icon is not aligned properly.

That's what that icon looks like! It's directly from remix and the [same way in our library](https://design-system.appsmith.com/?path=/story/design-system-icon--icon-story&args=name:checkbox-circle-fill;size:lg)

> Wrong height of the warning icon

In Zeplin, we should look at the outer container to determine the size of the icon. So, this size is correct.

shadabbuchh commented:
@tanvibhakta Below issues still exist:
**Default Toast:**

1. Design:

- No Box Shadow

**Toast with action:**

1. Design:

- No Box Shadow

Noticed additional issues after the fix:

- The icons for the default toast are not aligned with the text appropriately (like they are in `Toast with action`).
- The Toast with action does not disappear after we click on it.

> 3. Can we please add some description in Docs? Was it not added intentionally?

Any thoughts on the above?
Please refer to the below video for better clarity:
https://www.loom.com/share/db2576b33ac541c3b280174ea9e6877d

tanvibhakta commented:
> box shadow

My bad, have fixed and verified

> description

It's strange why the description is not being picked up from the doc strings. I've now added it explicitly! This adds to our maintenance overhead but I don't have the time to figure this out right now. 

> alignment 

So I checked - both toast and toast with action are ultimately the same component, so they are both center aligned! I believe it looks differently aligned because of the difference in text tricking our eyes - when I used the same text in both the stories it was clear to me that they were correctly aligned. I also think the feeling of the alignment being off is coming from the fact that the icons are 20px, but this is a call that design has made :/

> action on click 

so for a regular toast, we want it to disappear after we click on it. But if we allow that behavior when it has a button in it, the toast is unable to differentiate between a click to dismiss and a click for the button. so, we disable click to dismiss when there is a button, and we treat the button as the dismiss trigger. Therefore what you're seeing is the expected behavior (and I hope, the intuitive behavior when the toast appears in a layout as well!)

shadabbuchh commented:
All issues have been resolved. This is good to go.

shadabbuchh commented:
Something to look at a later stage:
If the Button in Toast has longer text, The UI breaks and the button looks ugly.
<img width="375" alt="Screenshot 2023-02-15 at 11 34 17 AM" src="https://user-images.githubusercontent.com/39921438/218946117-bef8b1b0-5ed9-4e0c-aecd-928020421c47.png">
<img width="1440" alt="Screenshot 2023-02-15 at 11 34 34 AM" src="https://user-images.githubusercontent.com/39921438/218946129-c5afb909-81dc-4613-b6b7-1ebe068c25b1.png">
